lib/orderbook/OrderBook.ts

```diff
@@ -250,106 +144,77 @@ class OrderBook extends EventEmitter {
-   * Add an order to an order map
-   * @returns false if an order with the same id already exist, otherwise true
-   */
-  private addOrder = (ordersMap: OrdersMap, order: orders.StampedOrder) => {
-    if (this.isOwnOrdersMap(ordersMap)) {
-      const { localId } = order as orders.StampedOwnOrder;
-      this.localIdMap.set(localId, order.id);
-    }
-
-    const orderMap = this.getOrderMap(ordersMap, order);
-    if (orderMap.has(order.id)) {
+  private removeOwnOrder = (orderId: string, pairId: string): boolean => {
+    const matchingEngine = this.matchingEngines.get(pairId);
+    if (!matchingEngine) {
+      this.logger.warn(`invalid pairId: ${pairId}`);
       return false;
-    } else {
-      orderMap.set(order.id, order);
-      return true;
     }
-  }
 
-  private removeOrder = (ordersMap: OrdersMap, orderId: string, pairId: string): boolean => {
-    if (!this.pairIds.includes(pairId)) {
-      throw errors.INVALID_PAIR_ID(pairId);
-    }
-    const orders = ordersMap.get(pairId);
-
-    if (orders) {
-      if (orders.buyOrders.has(orderId)) {
-        orders.buyOrders.delete(orderId);
-        return true;
-      } else if (orders.sellOrders.has(orderId)) {
-        orders.sellOrders.delete(orderId);
-        return true;
-      }
+    const order = matchingEngine.removeOwnOrder(orderId);
+    if (!order) {
+      this.logger.warn(`invalid orderId: ${pairId}`);
+      return false;
     }
 
-    return false;
+    this.localIdMap.delete(order.localId);
+    this.logger.debug(`order removed: ${JSON.stringify(orderId)}`);
+    return true;
   }
 
-  private getOrderMap = (ordersMap: OrdersMap, order: orders.StampedOrder): Map<string, orders.StampedOrder> => {
-    const orders = ordersMap.get(order.pairId);
-    if (!orders) {
-      throw errors.INVALID_PAIR_ID(order.pairId);
+  private removePeerOrder = (orderId: string, pairId: string, quantityToDecrease?: number): orders.StampedPeerOrder | undefined => {
+    const matchingEngine = this.matchingEngines.get(pairId);
+    if (!matchingEngine) {
+      this.logger.warn(`incoming order invalidation: invalid pairId (${pairId})`);
+      return;
     }
-    if (order.quantity > 0) {
-      return orders.buyOrders;
+    const order = matchingEngine.removePeerOrderQuantity(orderId, quantityToDecrease);
+    if (!order) {
+      this.logger.warn(`incoming order invalidation: invalid orderId (${orderId})`);
+      return;
     } else {
-      return orders.sellOrders;
+      assert(order.quantity === quantityToDecrease, 'order quantity must equal quantityToDecrease');
+      this.emit('peerOrder.invalidation', { orderId, pairId, quantity: order.quantity });
+      return order;
     }
   }
 
-  private isOwnOrdersMap = (ordersMap: OrdersMap) => {
-    return ordersMap === this.ownOrders;
+  private removePeerOrders = async (peer: Peer): Promise<void> => {
+    this.matchingEngines.forEach((matchingEngine) => {
+      const orders = matchingEngine.removePeerOrders(peer.nodePubKey!);
```

Comment:
We should check if the peer supports the pair of the matching engine first.

Reply:
Right.. I added a TODO, because it's the equivalent to the TODO of sending only supported pairs (line 225).